rust/lance/src/dataset/write/update.rs

```diff
@@ -0,0 +1,566 @@
+                Expr::Literal(value @ ScalarValue::List(_, _))
+                    if matches!(dest_type, DataType::FixedSizeList(_, _)) =>
+                {
+                    Expr::Literal(safe_coerce_scalar(&value, &dest_type).ok_or_else(|| {
+                        ArrowError::CastError(format!(
+                            "Failed to cast {} to {} during planning",
+                            value.data_type(),
+                            dest_type
+                        ))
+                    })?)
+                }
+                _ => expr
+                    .cast_to(&dest_type, &df_schema)
+                    .map_err(box_error)
+                    .context(InvalidInputSnafu)?,
+            };
+        }
+
+        // Optimize the expression. For example, this might apply the cast on
+        // literals.
+        let expr = planner
+            .optimize_expr(expr)
+            .map_err(box_error)
+            .context(InvalidInputSnafu)?;
+
+        self.updates.insert(column.as_ref().to_string(), expr);
+        Ok(self)
+    }
+
+    // TODO: should we support Expr for value instead of &str?
+    // pub fn set_expr(mut self, column: &str, value: Expr) -> Result<Self> { ... }
+
+    // TODO: set write params
+    // pub fn with_write_params(mut self, params: WriteParams) -> Self { ... }
+
+    pub fn build(self) -> Result<UpdateJob> {
+        let mut updates = HashMap::new();
+
+        let planner = Planner::new(Arc::new(self.dataset.schema().into()));
+
+        for (column, expr) in self.updates {
+            let physical_expr = planner.create_physical_expr(&expr)?;
+            updates.insert(column, physical_expr);
+        }
+
+        if updates.is_empty() {
+            return Err(Error::invalid_input("No updates provided", location!()));
+        }
+
+        let updates = Arc::new(updates);
+
+        Ok(UpdateJob {
+            dataset: self.dataset,
+            condition: self.condition,
+            updates,
+        })
+    }
+}
+
+// TODO: support distributed operation.
+
+#[derive(Debug, Clone)]
+pub struct UpdateJob {
+    dataset: Arc<Dataset>,
+    condition: Option<Expr>,
+    updates: Arc<HashMap<String, Arc<dyn PhysicalExpr>>>,
+}
+
+impl UpdateJob {
+    pub async fn execute(self) -> Result<Arc<Dataset>> {
+        let mut scanner = self.dataset.scan();
+        scanner.with_row_id();
+
+        if let Some(expr) = &self.condition {
+            scanner.filter_expr(expr.clone());
+        }
+
+        let stream = scanner.try_into_stream().await?.into();
+
+        // We keep track of seen row ids so we can delete them from the existing
+        // fragments.
+        let removed_row_ids = Arc::new(RwLock::new(RoaringTreemap::new()));
+        let stream = make_rowid_capture_stream(removed_row_ids.clone(), stream)?;
+
+        let schema = stream.schema();
+
+        let expected_schema = self.dataset.schema().into();
+        if schema.as_ref() != &expected_schema {
```

Comment:
Is this a debug assert?  How could this happen?

Reply:
Yeah this is a debug one. Hence why it's an internal error. Seems worth it though just in case we messed up somehow; don't want that data written to the table.